### PR TITLE
Cosigned: Webhook to ensure defined namespaces use signed images

### PR DESCRIPTION
**Pull request by cooktheryan**
Cosigned will be used to verify images are correctly signed. The signing is generated from tekton chains. 

This will help to close this https://github.com/AICoE/summit-2021-octo-keynote/issues/16

Signed-off-by: Ryan Cook <rcook@redhat.com>

### CODE REVIEW

**Comment by larsks** [comment]
It would be great if you could update the commit message to answer:

- What is cosigned?
- Why are we installing it?

Thanks!

**Comment by cooktheryan** [comment]
@larsks patched. Sorry about that

**Comment by larsks** [comment]
@cooktheryan the PR description is updated, but the commit message itself needs the same treatment.

**Comment by cooktheryan** [comment]
@larsks got it!

**Comment by larsks** [comment]
I added the explanatory text from the pr description to the commit message.

File: cosigned/base/clusterrole.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
```
Comment (HumairAK) [inline cosigned/base/clusterrole.yaml:3]: should go in: https://github.com/operate-first/apps/tree/master/cluster-scope/base/rbac.authorization.k8s.io 

and add to https://github.com/operate-first/apps/blob/master/cluster-scope/overlays/prod/moc/smaug/kustomization.yaml

File: cosigned/base/clusterrolebinding.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
```
Comment (HumairAK) [inline cosigned/base/clusterrolebinding.yaml:3]: should go in: https://github.com/operate-first/apps/tree/master/cluster-scope/base/rbac.authorization.k8s.io 

and add to https://github.com/operate-first/apps/blob/master/cluster-scope/overlays/prod/moc/smaug/kustomization.yaml

File: cosigned/base/mutatingwebhook.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+---
+apiVersion: admissionregistration.k8s.io/v1
+kind: MutatingWebhookConfiguration
```
Comment (HumairAK) [inline cosigned/base/mutatingwebhook.yaml:3]: cluster resource, should go in: 

https://github.com/operate-first/apps/tree/master/cluster-scope/base
(you'll need to create a new folder here, following the same format as the other resources, see readme [here](https://github.com/operate-first/apps/tree/master/cluster-scope) for more info)

File: cosigned/base/namespace.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+  labels:
+    kubernetes.io/metadata.name: cosign-system
+  name: cosign-system
```
Comment (HumairAK) [inline cosigned/base/namespace.yaml:6]: not needed, as this was added [here](https://github.com/operate-first/apps/blob/73fe9f7d5c790f1b3521cda67963ef96b9f620f8/cluster-scope/base/core/namespaces/cosigned/namespace.yaml)

File: cosigned/base/validatingwebhook.yaml [hunk 5]
```diff
@@ -0,0 +1,20 @@
+---
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingWebhookConfiguration
```
Comment (HumairAK) [inline cosigned/base/validatingwebhook.yaml:3]: same as the `mutatingwebhookconfiguration` this is cluster resource, should go in:

https://github.com/operate-first/apps/tree/master/cluster-scope/base

**Comment by cooktheryan** [comment]
sorry about that @HumairAK still trying to get my feet under me in the operate first envs

**Comment by HumairAK** [comment]
@cooktheryan no problem at all! It's not exactly obvious right off the bat, we totally get it!

**Comment by cooktheryan** [comment]
@HumairAK updated. I don't believe we will deploy this on Smaug but we will define it with Argo to deploy to remote clusters as we won't need the cosigned functionality in Smaug

File: cosigned/base/validatingwebhook.yaml (hunk 5 above)
Comment (harshad16) [reply cosigned/base/validatingwebhook.yaml:3]: @cooktheryan you have moved it to cluster-scope.
however forget to remove it from here

File: cosigned/base/kustomization.yaml [hunk 6]
```diff
@@ -0,0 +1,15 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+namespace: cosigned
+resources:
+- ../../cluster-scope/base/rbac.authorization.k8s.io/clusterrolebindings/cosigned
+- ../../cluster-scope/base/rbac.authorization.k8s.io/clusterroles/cosigned
```
Comment (harshad16) [inline cosigned/base/kustomization.yaml:6]: These needs to move to cluster side overlays.
like here: https://github.com/operate-first/apps/blob/d8add02ad6fd58a961ba404b6318c744ea255e19/cluster-scope/overlays/prod/moc/smaug/kustomization.yaml#L100

File: cosigned/base/kustomization.yaml [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+- configmap.yaml
+- deployment.yaml
+- ../../cluster-scope/base/cosigned
```
Comment (harshad16) [inline cosigned/base/kustomization.yaml:9]: Same about this need to be on cluster side.
https://github.com/operate-first/apps/blob/d8add02ad6fd58a961ba404b6318c744ea255e19/cluster-scope/overlays/prod/moc/smaug/kustomization.yaml#L100

File: cosigned/base/validatingwebhook.yaml (hunk 5 above)
Comment (cooktheryan) [reply cosigned/base/validatingwebhook.yaml]: @harshad16 apologies just sent in patch

File: cosigned/base/kustomization.yaml [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+- publicsecret.yaml
+- rolebinding.yaml
+- role.yaml
+- secret.yaml
+- serviceaccount.yaml
+- service.yaml
```
Comment (harshad16) [inline cosigned/base/kustomization.yaml:15]: need a new line after this

**Review by harshad16 — CHANGES_REQUESTED** [review]
some more minor changes needed

**Comment by harshad16** [comment]
@cooktheryan , do we want to test this on smaug or on some other cluster ?

**Comment by cooktheryan** [comment]
@harshad16 does this look right? I'm trying to make sure I understand the cluster side definition 
```
--- a/cosigned/base/kustomization.yaml
+++ b/cosigned/base/kustomization.yaml
@@ -2,14 +2,15 @@ apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 namespace: cosigned
 resources:
-- ../../cluster-scope/base/rbac.authorization.k8s.io/clusterrolebindings/cosigned
-- ../../cluster-scope/base/rbac.authorization.k8s.io/clusterroles/cosigned
+- ../../../../base/rbac.authorization.k8s.io/clusterrolebindings/cosigned
+- ../../../../base/rbac.authorization.k8s.io/clusterroles/cosigned
 - configmap.yaml
 - deployment.yaml
-- ../../cluster-scope/base/cosigned
+- ../../../../../../../../base/cosigned
 - publicsecret.yaml
 - rolebinding.yaml
 - role.yaml
 - secret.yaml
 - serviceaccount.yaml
-- service.yaml
\ No newline at end of file
+- service.yaml
+
```

**Comment by harshad16** [comment]
I'm trying to make sure I understand the cluster side definition
> 
> ```
> --- a/cosigned/base/kustomization.yaml
> +++ b/cosigned/base/kustomization.yaml
> @@ -2,14 +2,15 @@ apiVersion: kustomize.config.k8s.io/v1beta1
>  kind: Kustomization
>  namespace: cosigned
>  resources:
>  - configmap.yaml
>  - deployment.yaml
>  - publicsecret.yaml
>  - rolebinding.yaml
>  - role.yaml
>  - secret.yaml
>  - serviceaccount.yaml
> +- service.yaml
> +
> ```

what i was suggesting was, those cluster resource would go in the different Kustomization file 
like for smaug in here: [apps/cluster-scope/overlays/prod/moc/smaug/kustomization.yaml](https://github.com/operate-first/apps/blob/d8add02ad6fd58a961ba404b6318c744ea255e19/cluster-scope/overlays/prod/moc/smaug/kustomization.yaml#L100)

if we don't want it in smaug, then surely we can have you had already stated.

File: cosigned/base/kustomization.yaml (hunk 8 above)
Comment (harshad16) [inline cosigned/base/kustomization.yaml:15]:
```suggestion
- configmap.yaml
- deployment.yaml
- publicsecret.yaml
- rolebinding.yaml
- role.yaml
- secret.yaml
- serviceaccount.yaml
- service.yaml

```
The cluster resource don't need to be in here.

**Review by harshad16 — CHANGES_REQUESTED** [review]
all lgtm
just a final change

**Comment by codificat** [comment]
As prow is currently using a workaround for object storage that prevents the links to job details from working, copying the pre-commit error here from the pod's logs:

```
yamllint.................................................................Failed
- hook id: yamllint
- exit code: 1

cluster-scope/base/rbac.authorization.k8s.io/clusterroles/cosigned/clusterrole.yaml
  32:21     error    too many spaces inside brackets  (brackets)
  32:37     error    too many spaces inside brackets  (brackets)
```

**Comment by harshad16** [comment]
/lgtm
/approve
thanks :100:

**Review by harshad16 — APPROVED** [review]

**Pull request merged**